### PR TITLE
polkadot-parachain: Add omni-node variant with u64 block number

### PR DESCRIPTION
**Pull request by serban300**
Related to https://github.com/paritytech/polkadot-sdk/issues/4787

The main changes in this PR are the following:
- making the NodeSpec logic generic on the Block type
- adding an omni-node variant with u64 block number

Apart from this, the PR also moves some of the logic in `service.rs` to the `common` subfolder

The omni-node variant with u64 block number is not used yet. We have to either expose the option in the CLI or to read the block number from the chain spec somehow. Will do it in a future PR.

### CODE REVIEW

File: cumulus/polkadot-parachain/src/common/command.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/src/common/command.rs:74]: I've talked about this before, so possibly it is duplicate: 

we anticipate omni nodes to not have any benchmarking functionality, so I do encourage you to explore the state of this (e.g. how far we are from omni-bencher being self-sufficient) this and possibly NOT add code that we are going to remove later. 

Probably what I am suggesting today is not feasible, because we use the benchmark subcommand in many places in our CI. 

I am okay with you making this compromise now, but reminder that we need plan + tracking issue to remove benchmarking subcommand from here. The first step couple be keeping the subcommand but marking it as deprecated.

Comment (kianenigma) [reply cumulus/polkadot-parachain/src/common/command.rs:74]: Upon further thought, I questioned my own opinion. 

Keeping `benchmark`, `try-runtime` or `build-spec` subcommand is fine, as long as they don't require a hard dependency to the runtime. In the past they did rely on the runtime being linked as a dependency. 

Especially now that we have the omni-node library in place (https://github.com/paritytech/polkadot-sdk/pull/5288) and teams can abstract away from needing to define these commands in e.g. `struct Commands`, in fact it is beneficial to keep these subcommand in the omni-node, so that teams don't need multiple binaries to do their day-to-day work. 

Imagine how nicer it is to have, instead of: 

- `polkadot-parachain`
- `frame-omni-bencher`
- `chain-spec-builder` 

have one called `polkadot-omni-node`? 

It seems to me that we mainly separated these binaries _because_ we had leaky abstraction in the node side (e.g. `sturct Command` that had to be updated by each and every team to support a new subcommand).

TLDR; in light of https://github.com/paritytech/polkadot-sdk/pull/5288 which is a step in the direction of the original vision of #5, I don't see why things like omni-bencher and chain-spec-builder cannot be part of the omni-node again. 

@michalkucharczyk @ggwpez @bkchr wdyt?

**Review by kianenigma — COMMENTED** [review]
I personally think the best way to expose this is neither CLI nor chain-spec, and instead the pattern introduced in https://github.com/paritytech/polkadot-sdk/pull/5288, so I would wait for that to be merged first.

CLI and/or chain-spec and/or env variable and all indeed work, but the most meta and un-opinionated way is 5288.

File: cumulus/polkadot-parachain/src/common/command.rs
Comment (ggwpez) [reply cumulus/polkadot-parachain/src/common/command.rs:74]: > I don't see why things like omni-bencher and chain-spec-builder cannot be part of the omni-node again.

Yes the omni-node can definitely re-expose the commands with a flag. The advantage of the omni-bencher is faster compiler time and smaller binary for CI. Since building a complete node just for benchmarking is a bit of a waste.

Comment (serban300) [reply cumulus/polkadot-parachain/src/common/command.rs:74]: > I am okay with you making this compromise now, but reminder that we need plan + tracking issue to remove benchmarking subcommand from here. The first step couple be keeping the subcommand but marking it as deprecated.

Yes, agree. This is tracked here: https://github.com/paritytech/polkadot-sdk/issues/4966 . From what I udnerstand we still need https://github.com/paritytech/polkadot-sdk/pull/4405 to be merged, in order to remove the benchmarks from the node.

> Especially now that we have the omni-node library in place (https://github.com/paritytech/polkadot-sdk/pull/5288) and teams can abstract away from needing to define these commands in e.g. struct Commands, in fact it is beneficial to keep these subcommand in the omni-node, so that teams don't need multiple binaries to do their day-to-day work.

Keeping it also sounds good.

**Comment by serban300** [comment]
> I personally think the best way to expose this is neither CLI nor chain-spec, and instead the pattern introduced in #5288, so I would wait for that to be merged first.

Sounds good. I will come back to this after #5288 is merged

**Comment by serban300** [comment]
> > I personally think the best way to expose this is neither CLI nor chain-spec, and instead the pattern introduced in #5288, so I would wait for that to be merged first.
> 
> Sounds good. I will come back to this after #5288 is merged

Done. @kianenigma can you please take a look ?

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:59]: This match seems to grow exponentially in arms if coded as-is. Perhaps not now, but as we get more things to be parameterize-able, we should think about passing arguments down in a less verbose manner.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs:25]: In general would expect us to be more diligent in documenting this `lib` crate, as it is a public API that is meant to be used by externals. 

Perhaps try adding `#![deny(undocumented_pub_items)]` for this crate, and let it double check your choices?

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs:51]: Damn, this is unfortunate. Similar to the above, it _might_, depending on what is being made parameterize-able, explore exponentially. 

Although, not everything will affect this, right? 

 I wish we could pull a more blanket trick here and make the `node <-> runtime` communication not be _not-type-safe_. This degree of type-safety doesn't make sense for a node that is meant to work with various runtimes, and afaik is mainly the legacy of native runtime days.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs:30]:
```suggestion
pub type BlockOf<BlockNumber> = generic::Block<Header<BlockNumber>, UncheckedExtrinsic>;
```

nit, but usually we called such aliases in substrate like that.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs:31]: Questionable.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs
Comment (kianenigma) [inline cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs:34]: Sorry if this is duplicate, but can you elaborate why this is needed? 

Or you are onboard that it is only needed for the case of the hardcoded chains, not in the omni cases.

**Review by kianenigma — COMMENTED** [review]
Code looks broadly good, I will approve after another quick pass. 

Would be great if the second review is from a node-sdk expert. 

Beyond "the code compiles", what steps do you perform to make sure the scenario we have in mind in every step of the way is correct? 

we should work towards a number of network setups where we can quickly test relevant scenarios. Even if they are not tested in CI (for now), they are great help for us to ensure correctness of what we are doing. 

For example, as with this PR, I should expect to quickly run a zombienet network with a U64 block number, and see it work in action. 

This setup would then also help us verify some questions in the code. For example, at some point I had the guess that the types passed into `FakeRuntime` in `impl_runtime_apis` really don't matter AT ALL. As in, I think I got a node working with totally fake/dumb types. Having a simple test setup will help us test this kind of hypothesis, without risk breaking anything.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs
Comment (kianenigma) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs:34]: Okay you answered in https://github.com/paritytech/polkadot-sdk/pull/5269#discussion_r1713682255

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs
Comment (kianenigma) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs:25]: I would also hope the main `lib.rs` docs that go into `docs.rs` to be very extensive, and contain full examples of how the lib can be used. 

You can either do this here, or make an issue for it and assign me, I like doing such parts of the work :)

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/command.rs:84]: Yes, that's right. I think the only way to split it is to create multiple functions. For the moment I added one more `new_aura_node_spec` fn that takes that aura_id as a parameter, and here we just check the block number. I don't know if this split is super natural. But as we add more parameters, we can play and adjust this split.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/fake_runtime_api/mod.rs]: The `fake_runtime` is not public. The `mod` is private to the crate. But, good point, I added `#![deny(missing_docs)]` and there were some errors.

Added this issue for documentation: https://github.com/paritytech/polkadot-sdk/issues/5464 . I would prefer it to do it in a separate PR, since this one is already pretty verbose.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs]: I removed `CustomBlock`. Now it's just called `Block`. 

I saw this kind of naming in substrate but I think it was for things like: `HeaderOf<Block>` which is an alias for `Block::Header`. I wouldn't use `BlockOf` here since we're not accessing a property `Block` of something. We are referencing a Block that uses a certain number. But I don't know. I don't have strong preferences.

File: cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs
Comment (serban300) [reply cumulus/polkadot-parachain/polkadot-parachain-lib/src/common/types.rs]: Removed `CustomBlock` and left `Block<Number>`

**Comment by serban300** [comment]
> Beyond "the code compiles", what steps do you perform to make sure the scenario we have in mind in every step of the way is correct?
> 
> we should work towards a number of network setups where we can quickly test relevant scenarios. Even if they are not tested in CI (for now), they are great help for us to ensure correctness of what we are doing.

For this PR, personally I tested, asset hub polkadot, asset hub rococo and peregrine



> we should work towards a number of network setups where we can quickly test relevant scenarios. Even if they are not tested in CI (for now), they are great help for us to ensure correctness of what we are doing.
> 
> For example, as with this PR, I should expect to quickly run a zombienet network with a U64 block number, and see it work in action.

I think that in the CI there are zombienet tests that spawn most of the runtimes suported by polkadot-parachain. The bridges zombienet tests spawn for example, asset hub rococo, asset hub westend, bridge hub rococo, bridge hub westend

But yes, there is no test for a parachain with u64 block number.

I agree it would be good to have a test setup. But also, I think we should do it as part of a separate PR/issue

**Comment by bkchr** [comment]
> As in, I think I got a node working with totally fake/dumb types. Having a simple test setup will help us test this kind of hypothesis, without risk breaking anything.

This hypothesis is completely wrong. `u32` can not be decoded as `u64`, as `u64` requires 8 bytes versus 4 bytes for `u32`.

**Comment by kianenigma** [comment]
> > As in, I think I got a node working with totally fake/dumb types. Having a simple test setup will help us test this kind of hypothesis, without risk breaking anything.
> 
> This hypothesis is completely wrong. `u32` can not be decoded as `u64`, as `u64` requires 8 bytes versus 4 bytes for `u32`.

I see, so the `impl RuntimeApi for FakeRuntime` is not fully fully fake, the types that are provided there are still used in wasm execution, despite the fact that the code in the `fn` bodies are never called. I thought that part is purely a relic of the Native runtime and so on. 

But it makes sense, even in wasm we need some reference to decode the types I suppose.

**Review by kianenigma — APPROVED** [review]

**Pull request merged**
